### PR TITLE
feat(api): Add G-Code parsing

### PR DESCRIPTION
**Pull request by DerekMaggio**
# Overview

Add functionality for capturing and parsing G-Code.

# Changelog

* Added functionality to emulator log to file whenever a G-Code is sent 
* Added class to convert log file into a collection of GCode objects contained inside a GCodeProgram object
* Refactored `src/opentrons/hardware_control/emulation/parser.py` to expose parsing functionality for string creation. Not just commands
* Added unit tests for new code
* Changed from using a regex to parse G-Code to just simple string splitting
* Added tests for validating G-Code args
* Removed file defining expected results for test_g_code and instead just defined it in a list
* Instead of using logging now using mocking to extract G-Codes

# Review requests

Take a close look at get_gcode_function method inside of g_code and let me know if I missed edge cases

# Risk assessment

Logging could affect performance of emulator

### CODE REVIEW

File: api/src/opentrons/hardware_control/g_code_parsing/utils.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import re
+from typing import Dict, Type
+from opentrons.drivers.smoothie_drivers.driver_3_0 import GCODE as SMOOTHIE_GCODE
+
+
+WRITE_REGEX = re.compile(r"(.*?) \| (.*?) \|(.*?)$")
```
Comment (amitlissack) [inline api/src/opentrons/hardware_control/g_code_parsing/utils.py:6]: Is this regex needed in this file?

**Review by amitlissack — COMMENTED** [review]
A minor comment for now. 

This is really cool. We'll talk about some of the details and opportunities tomorrow.

File: api/src/opentrons/hardware_control/g_code_parsing/utils.py (hunk 1 above)
Comment (DerekMaggio) [reply api/src/opentrons/hardware_control/g_code_parsing/utils.py:6]: No. No I don't I will just change it to splitting on the pipe. Was using a more complex parsing method to start that was using regex. Didn't think not to use regex when I simplified it

Comment (DerekMaggio) [reply api/src/opentrons/hardware_control/g_code_parsing/utils.py:6]: Addressed this in my latest update

**Review by amitlissack — APPROVED** [review]
Nice job!

**Pull request merged**